enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,520 @@
+success/failure of that: re-balances the egress assignment to other nodes (thus
+avoiding packets being sent to a egress node which is currently un-reachable).
+Going forward, both network plugins will come to support traffic load-balancing
+and a more sophisticated route health detection. That means that the underlying
+networking solution (OVS in the case of both OVN-Kubernetes and openshift-sdn)
+will manage the route health check. It is worth noting that the "amount of
+re-balancing" done by the network plugin is expected to be reduced, since the
+network fabric will be able to load balance traffic between all egress nodes
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+##### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+**annotated approach**
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`network.openshift.io/node-cloud-ifaddr: {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}`
+
+This will in turn be read by the network plugin and used during the egress IP
+assignment, superseding any host level subnet information already parsed by the
+network plugin.
+
+**CRD**
+
+The cloud component will define a dedicated CRD that the network plugin will use
+to issue cloud egress IP requests. The network plugin will in-turn finalize its
+setup and set its final egress IP status pending the CR's status indicating a
+successful/unsuccessful cloud assignment. For example: if the cloud assignment
+should be unsuccessful; it will be up to the network plugin to notify the user
+of this, by setting its own status field for the egress IP assignment
+accordingly.
+
+The CRD will have `scope: Cluster` (since egress IP is cluster-scoped and node
+private IP assignment have no relation to namespaces) and have the following
+format:
+
+```yaml
+apiVersion: network.openshift.io/v1
+kind: CloudPrivateIPConfig
+metadata:
+  name: cloud-egressip
+spec:
+  items:
+  - node: nodeX
+    egressIP: 192.168.126.11
```

Comment:
All very fair questions, which tell me I should revisit this enhancement proposal. I wrote it with the intention that my reader would know the feature and have a bunch of implicit information already known, but I do feel like your questions are valid and could be clarified further. 

>  Any reason not to have one CloudPrivateIPConfig per node? 

Mainly because egress IP feature (as implemented by the network plugins) is cluster scoped, so this CRD needs to be so too. 

> what happens if you can't assign a single nodes egress, does that prevent others from reconciling? 

No, it should try adding every node's egress IP that it can and re-queue the ones it couldn't for the next reconciliation loop. It's not a "one fails, blocks all" type of scenario, more like "do as much as you can".  

> Does the name of the object matter, is it a singleton? 

The resource will be cluster scoped, so yes. 

>  What happens if you have two CloudPrivateIPConfig resources that define the same node?

That is not a problem as long as you have two different IPs. Ex:

```
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip-namepsace-1
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
---
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip-namespace-2
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.13
```
Is valid. This component needs validate that no conflicting information is provided, because:

```
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip-namepsace-1
spec:
  items:
  - node: nodeY
    egressIP: 192.168.126.11
---
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip-namespace-2
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
```

Is invalid, because that would break "IP address uniqueness"

Reply:
> > Any reason not to have one CloudPrivateIPConfig per node?

> Mainly because egress IP feature (as implemented by the network plugins) is cluster scoped, so this CRD needs to be so too.

To clarify, my suggestion was to still have the resource cluster scoped, but have a separate object per node.

> > Does the name of the object matter, is it a singleton?

> The resource will be cluster scoped, so yes.

Just because a resource is cluster scoped doesn't mean you can only have one. I could create two cluster scoped resources with different names (like nodes for example). It currently isn't defined how you will handle this scenario. One approach could be to block users from creating one unless it has a certain name and forcing only a single resource in that way. 

> Is invalid, because that would break "IP address uniqueness"

That's a very good point! (Which I don't think is obvious currently from the proposal). If your plan is to have validation so that every item in the list is unique, you should note that in the proposal, at the moment the structure of this data does not force that so I currently can't tell the difference in a list vs individual CR for this case. I think in both cases you could have some component that does this validation (a validating webhook is the only way I can really think to do this unless the controller does so)

Reply:
> I could create two cluster scoped resources with different names (like nodes for example). It currently isn't defined how you will handle this scenario.

So, as mentioned the only actual rule that must at all times be enforced is:

* Any IP can only be assigned to one node, and one node only, at any given moment in time. The reason for this is "IP address uniqueness"

A second rule can however be stated for the sake of clarity but should implicitly be covered by the controllers idempotency.  

* No IP can be assigned twice to the same node. 

There can be as many objects as possible with as many IPs requested for one node as necessary by the network plugin, as long as those rules are respected there should not be a problem.

I will state this in the enhancement proposal as I agree, it's lacking clarification. 

> I think in both cases you could have some component that does this validation (a validating webhook is the only way I can really think to do this unless the controller does so)

Right now, this is intended on being validated in the controller but a validating webhook is probably better though, agreed.